### PR TITLE
expand the possible inputs to `reversed`

### PR DESCRIPTION
**Pull request by tungol**
A bit experimental.

related to https://github.com/python/typeshed/issues/13218 and https://github.com/python/typeshed/issues/11645

### CODE REVIEW

**Comment by tungol** [comment]
I don't think this really does what we'd want because mypy doesn't handle `__new__` methods that return something other than a subclass of the type, and the new branch of the overload returns `reversed[T]` as far as mypy is concerned.

It's a tough spot to be in. We can expand what reversed will accept in it's constructor, but without mypy support it's only going to result in `reversed[T]`. If the actual return type is too divergent from that, that's an issue.

I'll check out the mypy-primer hits though to see what's going on and if they're actually being cleared in a reasonable way.

**Comment by tungol** [comment]
That said this *does* work as expected in pyright.

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -1730,11 +1730,16 @@ def pow(base: _SupportsSomeKindOfPow, exp: complex, mod: None = None) -> complex
 
 quit: _sitebuiltins.Quitter
 
+class _SupportsReversed(Protocol[_T_co]):
+    def __reversed__(self) -> _T_co: ...
+
 class reversed(Generic[_T]):
     @overload
     def __new__(cls, sequence: Reversible[_T], /) -> Iterator[_T]: ...  # type: ignore[misc]
```
Comment (srittau) [inline stdlib/builtins.pyi:1738]: I don't think we even need this overload. The second overload should include this one.

**Comment by srittau** [comment]
We could also use some test cases for this.

**Comment by tungol** [comment]
I don't think we can do this without mypy support, whether that's https://github.com/python/mypy/issues/15182 or special-casing for `reversed`.

For example, it's true that we don't need the first overload, but without that, and with the current state of mypy, we would get `reversed(list[int]) -> reversed[list[int]]` from mypy. Right now, the unpacking in `__new__(cls, sequence: Reversible[_T], /) -> Iterator[_T]` means we get `reversed[int]` from mypy.

We could, with more safety, expand this to accept `__reversed__ -> Iterable[T]`, but it's not ideal because mypy will still incorrectly infer `reversed[T]`, which is a full iterator.

So I don't really think it's worth worrying about test cases; this is busted and we shouldn't do it right now.

**Pull request closed without merging**